src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.cs

```diff
@@ -14,6 +18,73 @@ internal static void VerifyValidPath(string path, string argName)
             }
         }
 
+        internal static bool Copy(string sourcePath, string destinationPath, bool recursive,
+            bool skipExistingFiles, CancellationToken cancellationToken)
+        {
+            SearchOption searchOption = recursive ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly;
+            IEnumerable<string> directoryEnumeration;
+            bool success = true;
+
+            try
+            {
+                EnumerationOptions enumerationOptions = EnumerationOptions.Compatible;
+
+                // Compatible uses `IgnoreInaccessible = false`
+                enumerationOptions.IgnoreInaccessible = true;
+
+                if (recursive)
+                {
+                    enumerationOptions.RecurseSubdirectories = true;
+                }
+
+                directoryEnumeration = Directory.EnumerateDirectories(sourcePath, "*", enumerationOptions);
+            }
+            catch (IOException)
+            {
+                // Return false if Enumeration is not possible
+                return false;
+            }
+
+            foreach (string enumeratedDirectory in directoryEnumeration)
+            {
+                string newDirectoryPath = enumeratedDirectory.Replace(sourcePath, destinationPath);
+                Directory.CreateDirectory(newDirectoryPath);
+
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+
+            foreach (string enumeratedFile in Directory.GetFiles(sourcePath, "*.*", searchOption))
+            {
+                if (skipExistingFiles && File.Exists(enumeratedFile))
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    continue;
+                }
+
+                try
+                {
+                    CopyFile(enumeratedFile, enumeratedFile.Replace(sourcePath, destinationPath), true);
+                }
+                catch (IOException ex)
+                {
+                    switch (ex.HResult)
+                    {
+                        // success = false for read failures,
+                        // throw for everything else
+                        case Interop.Errors.ERROR_ACCESS_DENIED:
+                            success = false;
+                            break;
+                        default:
+                            throw;
+                    }
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+
+            return success;
+        }
+
```

Comment:
I doubt that this will result in better performance. [`IFileOperation::CopyItems`](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifileoperation-copyitems) is a COM interface method (I personally have [aversions](https://www.google.com/search?channel=nrow5&client=firefox-b-d&q=Why+is+COM+Interop+slow) to these). But I can create a benchmark for the Windows implementation as soon as I have time.

Whether [`cp`](https://man7.org/linux/man-pages/man1/cp.1.html) achieves a better performance I can't say so. However, I found the [source code](`https://github.com/coreutils/coreutils/blob/master/src/cp.c#L570-L781`) which might help us a bit.

@tmds can you maybe give us your advice because you have also dealt with Unix in another IO-PR?

Reply:
Also the command ``cp`` I think also works on Windows as well if I rember right. As for it's performance I do not know for sure.

Reply:
@AraHaan 
But what would be the best way to test this? I mean calling the `cp` command via BenchmarkDotNet I don't see as useful, because we have the whole process creation overhead, ...

Reply:
In the API review process it has been said (like I already commented here) that WinAPI implementation shouldn't be used. The video of it is still available.

Reply:
@Symbai I don't remember when that was said on API review but it makes sense after seeing that `Directory.Delete` uses `FileSystemEnumerable` on its implementation, thanks for pointing that out.

Reply:
> Directory.Delete uses FileSystemEnumerable on its implementation

Yes, that was added in https://github.com/dotnet/runtime/pull/59520, and it would be preferable to use it here too.

For deleting files, there is a need to start from the leaves because you can't copy an empty directory. That means we couldn't use `RecurseSubdirectories` on the enumerable. In this case, we don't have that restriction, and we can use a single enumerable with `RecurseSubdirectories` set.

Reply:
Perhaps new API #1953 (`OnDirectoryFinishedDelegate`) could be useful.

Reply:
@deeprobin can you change the implementation to use `FileSystemEnumerable` with `RecurseSubdirectories` set to `true`?

Reply:
@tmds I will do that in the days.

Reply:
@tmds Done. Could you please take a look at whether this corresponds to what you want?